<a href="https://colab.research.google.com/github/maritamayo25/Business-Analytics/blob/main/Reto_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Modelo Naive Bayes accidentes automovilisticos**

 0. Cargar las librerías de trabajo

In [ ]:
import numpy as np #Funciones matemáticas
import pandas as pd #Conexión con archivos de Excel

from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix #Metricas - como lo voy a evaluar

1. Carga de los datos

In [ ]:
nxl='/content/1. BD2_CarCrash_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB.dropna() #para evitar celdas vacías
XDB.head(100)

#Creamos el DataMart
#Creamos el DataMart
XD=XDB[['Weekday', 'ViolCat', 'ClearWeather', 'Month', 'Daylight']] #Variables Entrada
yd=XDB[['Highway']] #valor  deseado - aprenda el modelo



2.Se procede con la configuración y evaluación del modelo Naive Bayes


In [ ]:
len(XD)                #% de accidentados en la autopista o no
filas=np.where(yd==1)[0]
len(filas)
len(filas)/len(XD)

len(XD)
filas=np.where(yd==0)[0]
len(filas)
len(filas)/len(XD)



mnb=GaussianNB()
mnb.fit(XD,yd)

#Analizamos los resultados arrojados por el modelo
u=mnb.theta_ #son las medias
print("las medias de las variables son:\n", u)
sigma=np.sqrt (mnb.var_) #obtengo las desviaciones

#Se procede con el análisis de las variaciones de las variables
LS=u+sigma
print("los limites superiores para las variables son: \n", LS)
LI=u-sigma
print("los limites inferiores para las variables son: \n", LI)

print("los valores de las desviaciones son:\n", sigma)
#se procede con la evaluación del comportamiento del modelo
#prediciendo los valores de accidentes provocados en una autopista o no del valor deseado
ydp=mnb.predict(XD)

cm=confusion_matrix(yd,ydp)
print("la matriz de confusión es: \n",cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex=(VP+VN)/len(XD)
print("La exactitud del modelo es:",Ex)
# verdadero negativo, falso positivo
#falso negativo, verdadero positivo

Ter=(FP+FN)/len(XD)
print("La tasa de error del modelo es:", Ter)

Sen=VP/(VP+FN)
print("La sensibilidad del modelo es:", Sen)

Esp=VN/(VN+FP)
print("La especifidad del modelo es:", Esp)

Pres=VP/(VP+FP)
print("La precisión del modelo es:", Pres)

PredNeg=VN/(VN+FN)
print("El valor de predicción negativo del modelo es:", PredNeg)

#Para indicar
XSC=[[7,1,1,3,0]]
ydsc=mnb.predict(XSC)
print ("El ciudadano:", ydsc)






las medias de las variables son:
 [[3.88945752 6.7778915  0.85261003 5.52548618 0.6814739 ]
 [3.82417582 4.31538462 0.79450549 5.47417582 0.70934066]]
los limites superiores para las variables son: 
 [[ 5.81776877 10.41264957  1.20710429  8.79927955  1.14737865]
 [ 5.65218385  6.68332168  1.19856802  8.73632861  1.16340729]]
los limites inferiores para las variables son: 
 [[1.96114628 3.14313344 0.49811577 2.25169281 0.21556915]
 [1.9961678  1.94744755 0.39044297 2.21202304 0.25527403]]
los valores de las desviaciones son:
 [[1.92831125 3.63475806 0.35449426 3.27379337 0.46590475]
 [1.82800803 2.36793707 0.40406253 3.26215279 0.45406663]]
la matriz de confusión es: 
 [[4560  325]
 [1563  257]]
La exactitud del modelo es: 0.7184190902311708
La tasa de error del modelo es: 0.2815809097688292
La sensibilidad del modelo es: 0.1412087912087912
La especifidad del modelo es: 0.9334698055271239
La precisión del modelo es: 0.4415807560137457
El valor de predicción negativo del modelo es: 0.744

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


**Análisis de resultados**

Los resultados arrojados por el modelo propuesto frente al pronóstico de los accidentes automovilísticos provocados en una autopista mostro que la precisión alcanzo un porcentaje del 71% lo que indica que de un total de 100 accidentes automovilísticos el modelo tiende a equivocarse en 29 de ellos. Según la métrica de sensibilidad la cual alcanzó un valor de 14% frente a los accidentes automovilísticos que fueron provocados en la autopista y una especifidad de 93% frente a los accidentes automovilísticos **no** provocados en la autopista, lo que muestra que el modelo funciona mejor para Predecir si un accidente automovilístico **no** se provoco en la autopista, es importante tener en cuenta que el modelo arroja un informe de resultados en donde se evalua los accidentes automovilísticos provocados en la autopista o no para el condado de San Diego, para la predicción de los accidentes se tuvo en cuenta las siguientes variables: Día de la Semana, Nivel de Impacto (ViolCat), Mes, Cantidad de Luz Diurna, y Cielo Despejado, ya que estos factores contribuyen a los accidentes automovilísticos. Como información adicional la base de datos posee un total de 6705 accidentes automovilísticos, el 27% de ellos fueron provocados en la autopista y el 72,8% no fueron provocados en la autopista.